### ISSUE TITLE
Add support for 7z/LZMA1/LZMA2 to System.IO.Compression

### ISSUE THREAD

**Issue by joshfree** [dotnet/runtime#1542]
http://7-zip.org/sdk.html

**Comment by jjxtra** [comment]
LZ4 would be nice too, there's already nuget in pure C# with friendly licensing https://github.com/MiloszKrajewski/K4os.Compression.LZ4.

ZStandard is also interesting: https://github.com/bp74/Zstandard.Net, but it is not in pure C# yet.

**Comment by MrM40** [comment]
Really? That would be a surprise if W11 added native support for those formats. I think rar is a commercial non-open-source format. I guess 7zip is open-source, but it will still be a surprise if MS will implement it natively.  What happens when Igor Pavlov make changes to 7zip? I don't think MS will go down that rapid hole. Just thinking out loud.
Do you have a link?

**Comment by 2ji3150** [comment]
> Do you have a link?
https://www.neowin.net/news/opening-rar-files-natively-in-windows-11-is-coming-and-people-online-are-going-crazy-over-it/
Yap. Seems MSFT is going to use [libarchive](https://github.com/libarchive/libarchive)

**Comment by Symbai** [comment]
The Windows 11 update which supports RAR and 7zip has been released. How/when can we use that in .NET ?

**Comment by RokeJulianLockhart** [comment]
https://github.com/dotnet/runtime/discussions/92763#discussion-5681130

**Comment by KieranDevvs** [comment]
Seeing as .NET 8 has just been released and planning for the next version is taking place, is there any chance we could see this prioritised for .NET 9?

**Comment by 0xced** [comment]
> [!NOTE]  
> I'm a few weeks late for this [joke](https://arstechnica.com/security/2024/04/what-we-know-about-the-xz-utils-backdoor-that-almost-infected-the-world/) but let's post it anyway…

In the meantime I think someone should write a dotnet wrapper around [XZ Utils](https://github.com/tukaani-project/xz).

I heard version 5.6.1 is pretty fast thanks to @JiaT75 optimisations. 😂

**Comment by Mrgaton** [comment]
I need this so hard on .NET 8

**Comment by KieranDevvs** [comment]
@Mrgaton Agreed, there are no good LZMA/2 libraries and I've has several instances where it would have been very nice to have one.

**Comment by EatonZ** [comment]
Hello, I'm guessing the window has passed for this to have a chance at being included in .NET 9 later this year?

**Comment by KieranDevvs** [comment]
> Hello, I'm guessing the window has passed for this to have a chance at being included in .NET 9 later this year?

Yeah there's zero chance of that happening unfortunately.

A better question would be, how much customer engagement needs to happen before this can be prioritised. I'm not trying to be demanding or arrogant and I understand each release only has so much capacity and there are other features the .NET team think is more worth while, but this issue is almost 10 years old now and its 4th most top thumbed up issue on the entire repository. We're told that thumbing up and engaging with issues on GitHub is the best way to see issues/features get looked at. I feel like if that statement is true then this issue should have been picked up a long time ago? 

Can someone from the .NET team help me understand what specifically needs to happen to get this prioritised in the next available release? Is it more thumbs up? Is it more reports of customers use cases being blocked because LZMA isn't available?

**Comment by armoiredu44** [comment]
> http://7-zip.org/sdk.html

bro it's a pain to use, I can't find documentation anywhere

**Comment by Mrgaton** [comment]
> > http://7-zip.org/sdk.html
> 
> bro it's a pain to use, I can't find documentation anywhere

Completely agree

**Comment by somegenericdev** [comment]
> > http://7-zip.org/sdk.html
> 
> bro it's a pain to use, I can't find documentation anywhere

Other than being a pain to use, it has a hard limit of 2GB. If you try to compress a 2GB+ file, it will fail. Same if you try to decompress a file that was once 2GB+.

**Comment by KieranDevvs** [comment]
Glad to see this is being planned for .NET 11 👍

**Comment by Mrgaton** [comment]
So exited to see full implementation of this, 7zip format will be also added to make and read archives ?

**Comment by PaulusParssinen** [comment]
I was overjoyed to hear LZMA potentially being prioritized in todays .NET API Review! 🤞 (https://www.youtube.com/live/NzqIFvMFxWg?t=1357)

**Comment by kevincathcart-cas** [comment]
That makes sense. Exposing XZ and (separately) raw support for LZMA1/2 seems reasonable. Given the latter as a primitive, user code can easily support .lzma or microlzma, as those are merely headers added in front of a raw LZMA1 stream, and that covers all formats the underlying liblzma can create.

As for this proposal, what is your planning for the allow concatenation flag? The XZ specs says it is required for supporting XZ as a standalone file format, but that the typical expectation is that is not enabled when XZ is used as a stream inside a larger format. .NET mostly don't have decompression options like that, so presumably it will just be on all the time?

**Comment by bartonjs** [comment]
[Video](https://www.youtube.com/watch?v=L0LeEmrb9I8&t=0h13m3s)

Looks good as proposed

```csharp
namespace System.IO.Compression
{
  public enum XzChecksumType
  {
      None = 0,
      Crc32 = 1,
      Crc64 = 2,
      Sha256 = 3,
  }
  public sealed partial class XzCompressionOptions
  {
      public XzCompressionOptions() { }
      public static int DefaultQuality { get { throw null; } }
      public static int DefaultWindowLog { get { throw null; } }
      public static int MaxQuality { get { throw null; } }
      public static int MaxWindowLog { get { throw null; } }
      public static int MinQuality { get { throw null; } }
      public static int MinWindowLog { get { throw null; } }
      public System.IO.Compression.XzChecksumType Checksum { get { throw null; } set { } }
      public int WindowLog { get { throw null; } set { } }
      public bool EnableExtremeMode { get { throw null; } set { } }
      public int Quality { get { throw null; } set { } }
  }
  public sealed partial class XzDecoder : System.IDisposable
  {
      public XzDecoder() { }
      public XzDecoder(int maxWindowLog) { }
      public System.Buffers.OperationStatus Decompress(System.ReadOnlySpan<byte> source, System.Span<byte> destination, out int bytesConsumed, out int bytesWritten) { throw null; }
      public void Dispose() { }
      public static bool TryDecompress(System.ReadOnlySpan<byte> source, System.Span<byte> destination, out int bytesWritten) { throw null; }
  }
  public sealed partial class XzEncoder : System.IDisposable
  {
      public XzEncoder() { }
      public XzEncoder(int quality) { }
      public XzEncoder(int quality, int windowLog) { }
      public XzEncoder(System.IO.Compression.XzCompressionOptions compressionOptions) { }
      public System.Buffers.OperationStatus Compress(System.ReadOnlySpan<byte> source, System.Span<byte> destination, out int bytesConsumed, out int bytesWritten, bool isFinalBlock) { throw null; }
      public void Dispose() { }
      public System.Buffers.OperationStatus Flush(System.Span<byte> destination, out int bytesWritten) { throw null; }
      public static long GetMaxCompressedLength(long inputLength) { throw null; }
      public static bool TryCompress(System.ReadOnlySpan<byte> source, System.Span<byte> destination, out int bytesWritten) { throw null; }
      public static bool TryCompress(System.ReadOnlySpan<byte> source, System.Span<byte> destination, out int bytesWritten, int quality) { throw null; }
      public static bool TryCompress(System.ReadOnlySpan<byte> source, System.Span<byte> destination, out int bytesWritten, int quality, int windowLog) { throw null; }
  }
  public sealed partial class XzStream : System.IO.Stream
  {
      public XzStream(System.IO.Stream stream, System.IO.Compression.CompressionLevel compressionLevel, bool leaveOpen = false) { }
      public XzStream(System.IO.Stream stream, System.IO.Compression.CompressionMode mode, bool leaveOpen = false) { }
      public XzStream(System.IO.Stream stream, System.IO.Compression.XzCompressionOptions compressionOptions, bool leaveOpen = false) { }
      public System.IO.Stream BaseStream { get { throw null; } }
  }
}  
```

**Comment by rzikm** [comment]
After long and thorough discussion, we decided not to move forward with this API in .NET 11. There are several reasons, including:
- We already added Zstandard, which offers similar performance. The few cases where LZMA outperforms Zstandard (e.g. maximum compression with unlimited computation time) don't seem to be worth the cost
- by extension, same goes for 7z format, similar compression results as 7z can be achieved using .tar.zst
- many XZ/LZMA users (e.g. lots of package managers in Linux distros) have migrated to Zstandard which offers greater decompression speeds for comparable compression ratio
- future maintenance cost concerns (risk seems to be higher than zstd)

We will keep the issue open for the time being. We are open to feedback and discussion about our points above. If we missed anything important, please let us know.

**Comment by KieranDevvs** [comment]
> After long and thorough discussion, we decided not to move forward with this API in .NET 11. There are several reasons, including:
> 
> * We already added Zstandard, which offers similar performance. The few cases where LZMA outperforms Zstandard (e.g. maximum compression with unlimited computation time) don't seem to be worth the cost
> * by extension, same goes for 7z format, similar compression results as 7z can be achieved using .tar.zst
> * many XZ/LZMA users (e.g. lots of package managers in Linux distros) have migrated to Zstandard which offers greater decompression speeds for comparable compression ratio
> * future maintenance cost concerns (risk seems to be higher than zstd)
> 
> We will keep the issue open for the time being. We are open to feedback and discussion about our points above. If we missed anything important, please let us know.

Rather disappointing considering there has been a **TON** of community feedback for this. I don't really understand the argument of "we already have this algorithm that's different but does the same thing" when many of us have LZMA files that we don't control but need to consume and are having to rely on custom wrappers around other native implementations and some of us are just out right blocked from doing anything with LZMA because of this.

**Comment by karelz** [comment]
@KieranDevvs we understand your disappointment.
However, it is not clear to us, why do you think that our implementation would be any better than your custom wrappers around native implementation. We planned to pick one community native implementation and embed it anyway. What is preventing you from doing exactly that, if you need support specifically for LZMA?
Our main concern is that LZMA does not seem to be the way forward, therefore making everyone pay the penalty of larger size of embedded native implementation + the associated ongoing maintenance cost do not sound like a good investment for .NET.
Are we missing something here?

**Comment by jjxtra** [comment]
What about an "official" nuget package that isn't part of the core .NET offering?

**Comment by karelz** [comment]
@jjxtra that solves size angle, but has same (if not a bit larger) maintenance cost.
For something that is being replaced by Zstd, we are not sure it is well spend time/effort/energy for us. We believe we can add more value elsewhere in the repo, where there are no alternatives / workarounds for community.

**Comment by KieranDevvs** [comment]
> [@KieranDevvs](https://github.com/KieranDevvs) we understand your disappointment. However, it is not clear to us, why do you think that our implementation would be any better than your custom wrappers around native implementation. We planned to pick one community native implementation and embed it anyway. What is preventing you from doing exactly that, if you need support specifically for LZMA? Our main concern is that LZMA does not seem to be the way forward, therefore making everyone pay the penalty of larger size of embedded native implementation + the associated ongoing maintenance cost do not sound like a good investment for .NET. Are we missing something here?

Its not that wrapping a native implementation is in any way bad, its that you're asking everyone to write their own (which isnt a trivial task) for something which I consider a basic (from the consumer perspective, not the implementation perspective) IO function. Now maybe that's not necessarily bad either but you have to look at the community support for this and ask yourself why the landscape is so barren even though there are millions if not billions of LZMA archives out there and the standard has existed for almost 30 years and there are tons of people trying to all do the same thing, yet failing. From my experience its because the documentation is extremely lacking and the API surfaces usually aren't great to consume, and/or there's a performance issue.

As @jixtra suggested, the size concern can be mitigated via an isolated NuGet package, and while I understand your concern about having to support a library that isnt the latest and greatest thing, it doesn't sit right with me for a few reasons:

1) Yes LZMA may or may not be the way forward, but that is irrelevant IMO. The metrics that I think you should be concerned with are the community need for this (which has been proven to be desirable and high on many fronts (this GitHub issue, StackOverflow questions, and many failed community NuGet packages). Like I have previously mentioned, many people end up in a situation needing to read and deflate an LZMA archive or stream and they don't own the data and they cant change the archival format.

2) Supporting LZMA isnt like supporting a dying or dead technology, LZMA is still very much in use all over the planet and while something like zstandard can offer similar ratios, that doesn't change the fact that LZMA is still widely used and will be for years if not decades to come, and I don't see any evidence of that number declining just because an alternative exists. 

3) This might be my ignorance speaking so forgive me if I'm wrong, but how much support are you expecting to really be giving to this if it were to be added to the .NET ecosystem? The LZMA algorithm is extremely stable, and the 7zip implementation seems to get security updates every few months or so and there are on average between 1 and 3 on demand CVE's to patch which will be done by the native implementation provider (I'm assuming this will be 7zips). 

4) As I previously touched on, there is no easy way to consume these files in .NET. This proposal brings a clear OOB unified way on doing this.

And finally, why have we spent an entire decade requesting this to be added, gone through API design and approval and then at the last minute canned it all because someone suddenly doesn't like the idea of supporting it. Considering the .NET team keep asking people to give feedback and engagement around the things they would like to see prioritised, this seems like a huge slap in the face and extremely discouraging to engage with anything else.

I understand that there are pros and cons to this, but in my opinion the pros outweigh the cons quite significantly.

**Comment by karelz** [comment]
> at the last minute canned it all because someone suddenly doesn't like the idea of supporting it. Considering the .NET team keep asking people to give feedback and engagement around the things they would like to see prioritised, this seems like a huge slap in the face and extremely discouraging to engage with anything else.

I think this speaks actually in the opposite - the fact we cared enough to go with the API proposal, implementation, etc. (albeit after long time) demonstrates we DO CARE about community feedback and we take it seriously. We use it for prioritization.
There was not a whim of one person "killing it". Some (valid) concerns were raised in last phases of reviews about sustainability of our approach and we realized the whole effort is rather non-productive for .NET team to do based on the forward looking approach.
We believe that maintenance of the package is a high burden which we will regret later on. Obviously, we are very careful about that.
We are still open to change our mind, if there is very strong signal from the community, but the above is our current thinking.

> the standard has existed for almost 30 years

Technically speaking, it does not have RFC or ISO standard. The best we could find is a draft inside LZMA SDK. From practical purpose, it might be sufficient. But it is yet another small difference compared to Zstd.

If anyone is interested, we would be happy to contribute our API design and implementation we have to any OSS project. That way perhaps most of the asks above can be satisfied?
Would that be viable middle ground?
